extensions/sources/filebase/src/main/java/io/smallrye/ext/config/source/filebase/AbstractUrlBasedSource.java

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright 2019 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.smallrye.ext.config.source.filebase;
+
+import io.smallrye.ext.config.events.ChangeEventNotifier;
+import io.smallrye.ext.config.source.base.EnabledConfigSource;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Level;
+import lombok.extern.java.Log;
+
+
+/**
+ * URL Based property files
+ * 
+ * Load some file from a file and convert to properties.
+ * @author <a href="mailto:phillip.kruger@redhat.com">Phillip Kruger</a>
+ */
+@Log
+public abstract class AbstractUrlBasedSource extends EnabledConfigSource implements Reloadable {
+    private final LinkedHashMap<URL,Map<String, String>> propertiesMap = new LinkedHashMap<>();
+    private final Map<String, String> properties = new HashMap<>();
+    private String urlInputString;
+    private String keySeparator;
+    private boolean pollForChanges;
+    private int pollInterval;
+    private boolean notifyOnChanges;
+    
+    private FileResourceWatcher fileResourceWatcher = null;
+    private WebResourceWatcher webResourceWatcher = null;
+    
+    public AbstractUrlBasedSource(){
+        init();
+    }
+    
+    private void init(){
+        String ext = getFileExtension();
+        log.log(Level.INFO, "Loading [{0}] MicroProfile ConfigSource", ext);
+        this.keySeparator = loadPropertyKeySeparator();
+        this.notifyOnChanges = loadNotifyOnChanges();
+        this.pollForChanges = loadPollForChanges();
+        this.pollInterval = loadPollInterval();
+        this.urlInputString = loadUrlPath();
+        this.loadUrls(urlInputString);
+        
+        super.initOrdinal(500); 
+    }
+    
+    
+    @Override
+    protected Map<String, String> getPropertiesIfEnabled() {
+        return this.properties;
+    }
+
+    @Override
+    public String getValue(String key) {
+        // in case we are about to configure ourselves we simply ignore that key
```

Comment:
What happens if a user reads a property that starts with the configured prefix?  Stack overflow?

Reply:
That prefix is used to configure this specific config source, and a unconfigured config source can not configure itself. Ja ?